Ser hlslattributes patch

simoll opened the pull request:
Lowering for `HitObject::GetAttributes<T>()`

Specification: https://github.com/microsoft/hlsl-specs/blob/main/proposals/0027-shader-execution-reordering.md
DXC SER implementation tracker:: https://github.com/microsoft/DirectXShaderCompiler/issues/7214

simoll commented:
This PR is based on #7262 diff and will be updated once #7262 has been merged

jenatali commented on tools/clang/test/SemaHLSL/hlsl/objects/HitObject/hitobject_attributes.hlsl:
```diff
@@ -0,0 +1,28 @@
+// AST-NEXT: | | |   |-HLSLIntrinsicAttr {{[^ ]+}} <<invalid sloc>> Implicit "op" "" 364
+// AST-NEXT: | | |   `-AvailabilityAttr {{[^ ]+}} <<invalid sloc>> Implicit  6.9 0 0 ""
+
+// FCGL: %{{[^ ]+}} = call %struct.CustomAttrs* @"dx.hl.op..%struct.CustomAttrs* (i32, %dx.types.HitObject*)"(i32 364, %dx.types.HitObject* %{{[^ ]+}})
```
+1

If we do this change, we should consider revisiting the semantics of the DXIL op (though the signature is fine). If the structure is declared and the user has a chance to initialize it before passing it to the `GetAttributes` function, then we have two options for how `GetAttributes` should behave in the case of a hit object that doesn't store attributes (e.g. a no-op):
1. As it's currently spec'd, it must produce a zero-filled structure. We could maintain these semantics if we'd like.
2. We could spec that `GetAttributes` doesn't modify the structure if the hit object doesn't encode a procedural primitive hit.

I don't feel too strongly one way or another but I'd like to see a reason discussed in the spec for why one is chosen over the other.

tex3d replied:
I filed an issue here: https://github.com/microsoft/hlsl-specs/issues/495.

simoll replied:
Considering this resolved as far as this PR is concerned. Followup issue against the spec has been filed for follow up discussions.

llvm-beanz commented:
@simoll, one of the introduced tests is failing, can you fix that up?

simoll commented:
> @simoll, one of the introduced tests is failing, can you fix that up?

Fixed it. A diagnostic signature changed here while the traceinvoke PR added another instance of that diagnostic. Merged main and fixed that use too.
